src/libAtomVM/nifs.c

```diff
@@ -1284,14 +1284,43 @@ static term nif_erlang_spawn(Context *ctx, int argc, term argv[])
 static term nif_erlang_send_2(Context *ctx, int argc, term argv[])
 {
     UNUSED(argc);
+    term target = argv[0];
+    GlobalContext *glb = ctx->global;
+    int32_t receiver_id;
+
+    if (term_is_pid(target)) {
+        receiver_id = term_to_local_process_id(target);
+
+        globalcontext_send_message(glb, receiver_id, argv[1]);
+
+    } else if (term_is_atom(target)) {
+        struct ListHead *dummy = synclist_rdlock(&glb->processes_table);
+        UNUSED(dummy);
+        Context *p = NULL;
+        int atom_index = term_to_atom_index(target);
+
+        int target_id = globalcontext_get_registered_process(glb, atom_index);
+        if (UNLIKELY(target_id == 0)) {
+            goto badarg_error;
+        }
 
-    term pid_term = argv[0];
-    VALIDATE_VALUE(pid_term, term_is_pid);
+        p = globalcontext_get_process_nolock(glb, target_id);
+        if (IS_NULL_PTR(p)) {
+            goto badarg_error;
+        }
 
-    int local_process_id = term_to_local_process_id(pid_term);
-    globalcontext_send_message(ctx->global, local_process_id, argv[1]);
+        receiver_id = p->process_id;
+        globalcontext_send_message_nolock(glb, receiver_id, argv[1]);
+        synclist_unlock(&glb->processes_table);
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
 
     return argv[1];
+
+    badarg_error:
```

Comment:
just a last nitpicking moment: since we handle the error in this way just in case of `term_is_atom(...)`  just in 2 different places, I'd rather move back this code in those specific situations.
e.g.
```c
        if (UNLIKELY(target_id == 0)) {
            synclist_unlock(&glb->processes_table);
            RAISE_ERROR(BADARG_ATOM);
        }
```

Reply:
I don't think this is nitpicking. I almost submitted that change, but hesitated to change too much more the the requests you made. I am more than happy to fix this.